aws_lambda_powertools/utilities/idempotency/persistence/base.py

```diff
@@ -0,0 +1,432 @@
+                raise IdempotencyValidationerror("Payload does not match stored record for this event key")
+
+    def _get_expiry_timestamp(self) -> int:
+        """
+
+        Returns
+        -------
+        int
+            unix timestamp of expiry date for idempotency record
+
+        """
+        now = datetime.datetime.now()
+        period = datetime.timedelta(seconds=self.expires_after)
+        return int((now + period).timestamp())
+
+    def _save_to_cache(self, data_record: DataRecord):
+        self._cache[data_record.idempotency_key] = data_record
+
+    def _retrieve_from_cache(self, idempotency_key: str):
+        cached_record = self._cache.get(idempotency_key)
+        if cached_record:
+            if cached_record.is_expired:
+                logger.debug(f"Removing expired local cache record for idempotency key: {idempotency_key}")
+                self._delete_from_cache(idempotency_key)
+            else:
+                return cached_record
+
+    def _delete_from_cache(self, idempotency_key: str):
+        del self._cache[idempotency_key]
+
+    def save_success(self, event: Dict[str, Any], result: dict) -> None:
+        """
+        Save record of function's execution completing succesfully
+
+        Parameters
+        ----------
+        event: Dict[str, Any]
+            Lambda event
+        result: dict
+            The response from lambda handler
+        """
+        response_data = json.dumps(result)
+
+        data_record = DataRecord(
+            idempotency_key=self._get_hashed_idempotency_key(event),
+            status=STATUS_CONSTANTS["COMPLETED"],
+            expiry_timestamp=self._get_expiry_timestamp(),
+            response_data=response_data,
+            payload_hash=self._get_hashed_payload(event),
+        )
+        logger.debug(
+            f"Lambda successfully executed. Saving record to persistence store with "
+            f"idempotency key: {data_record.idempotency_key}"
+        )
+        self._update_record(data_record=data_record)
+
+        if self.use_local_cache:
+            self._save_to_cache(data_record)
+
+    def save_inprogress(self, event: Dict[str, Any]) -> None:
+        """
+        Save record of function's execution being in progress
+
+        Parameters
+        ----------
+        event: Dict[str, Any]
+            Lambda event
+        """
+        data_record = DataRecord(
+            idempotency_key=self._get_hashed_idempotency_key(event),
+            status=STATUS_CONSTANTS["INPROGRESS"],
+            expiry_timestamp=self._get_expiry_timestamp(),
+            payload_hash=self._get_hashed_payload(event),
+        )
+
+        logger.debug(f"Saving in progress record for idempotency key: {data_record.idempotency_key}")
+
+        if self.use_local_cache:
+            record = self._retrieve_from_cache(idempotency_key=data_record.idempotency_key)
+            if record:
+                raise ItemAlreadyExistsError
```

Comment:
Looks like this exception is only raised if use_local_cache is true. I didn't find this exception being raised in _put_record

Reply:
`_put_record` does raise the same error. I'll clarify in the ABC abstractmethod docstring though.

https://github.com/awslabs/aws-lambda-powertools-python/blob/c4856fdd0386460b329711907cf4b5b8466ba31e/aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py#L124